torch/_inductor/comms.py

```diff
@@ -342,6 +342,148 @@ def reorder_compute_and_comm_for_overlap(
     return order
 
 
+def remove_fsdp2_unsharded_param_graph_input_usage(graph: torch.fx.Graph):
+    """
+    This FX graph pass replaces uses of FSDP2 unsharded params with their corresponding
+    graph intermediates that were fsdp.copy_ into the unsharded params in the original graph.
+
+    NOTE: Can only apply this pass to FSDP2 unsharded params that are resharded (i.e. resized to 0)
+    at the end of graph.
+    """
+    node_list = list(graph.nodes)
+
+    # Find all graph inputs and their resize counts
+    graph_input_to_resized_to_full_node_idxes = defaultdict(list)
+    graph_input_to_resized_to_0_node_idxes = defaultdict(list)
+    for idx, node in enumerate(node_list):
+        if (
+            node.op == "call_function"
+            and node.target == torch.ops.inductor.resize_storage_bytes_.default
+            and node.args[0].op == "placeholder"
+        ):
+            graph_input = node.args[0]
+            new_size = node.args[1]
+            if new_size > 0:
+                graph_input_to_resized_to_full_node_idxes[graph_input].append(idx)
+            else:
+                graph_input_to_resized_to_0_node_idxes[graph_input].append(idx)
+
+    # Find all eligible unsharded params and their corresponding graph intermediates.
+    unsharded_param_to_fsdp_copy_node_idxes = defaultdict(list)
+    for idx, node in enumerate(node_list):
+        if node.op == "call_function" and node.target == torch.ops.fsdp.copy_.default:
+            fsdp_copy_node = node
+            unsharded_param = node.args[0]
+            assert (
+                unsharded_param.op == "placeholder"
+            ), "Assumed all FSDP2 `unsharded_param`s to be graph input, but it's not true!"
+            # An unsharded param is eligible for this pass only if it's resized to 0 at the end of graph.
+            num_resize_to_full = len(
+                graph_input_to_resized_to_full_node_idxes[unsharded_param]
+            )
+            num_resize_to_0 = len(
+                graph_input_to_resized_to_0_node_idxes[unsharded_param]
+            )
+            last_resize_to_full_node_idx = (
+                graph_input_to_resized_to_full_node_idxes[unsharded_param][-1]
```

Comment:
it seems like you never actually use the full list (we only ever grab the last element in it). If that's true then can't we just do away with the list completely, and only ever track the "last resize up/down applied to a given node"?

Although I think I'm still a bit confused: if there is a sequence of multiple resize ups/downs, won't there be two separate intermediate nodes that we need to use in different parts of the graph?

Reply:
> if there is a sequence of multiple resize ups/downs, won't there be two separate intermediate nodes that we need to use in different parts of the graph?

yes this happens when there is layer reuse: https://github.com/pytorch/pytorch/pull/133730#discussion_r1752407254. Updated check here to make it more clear